File: elevator-system/go/elevator.go
package main

import (
	"fmt"
	"math"
	"sync"
	"time"
)

// Elevator represents an individual elevator in the system
type Elevator struct {
	ID               string
	CurrentFloor     int
	CurrentDirection Direction
	Capacity         int
	CurrentLoad      int

	// Thread safety for adding/removing requests concurrently
	mu sync.RWMutex

	// Pending stops for the elevator (using maps as native Sets to ensure unique stops)
	UpStops   map[int]bool
	DownStops map[int]bool

	// Dedicated requests tracking passenger boarding/alighting
	Requests []*Request
}

func NewElevator(id string, capacity int) *Elevator {
	el := &Elevator{
		ID:               id,
		CurrentFloor:     0,
		CurrentDirection: IDLE,
		Capacity:         capacity,
		CurrentLoad:      0,
		UpStops:          make(map[int]bool),
		DownStops:        make(map[int]bool),
		Requests:         make([]*Request, 0),
	}
	// Start spinning the elevator's internal service loop concurrently
	go el.run()
	return el
}

func (e *Elevator) AddRequest(req *Request) error {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.CurrentLoad+req.Passengers > e.Capacity {
		return fmt.Errorf("elevator %s capacity exceeded", e.ID)
	}

	e.Requests = append(e.Requests, req)
	e.CurrentLoad += req.Passengers

	// Register source floor mathematically
	if req.SourceFloor > e.CurrentFloor {
		e.UpStops[req.SourceFloor] = true
	} else if req.SourceFloor < e.CurrentFloor {
		e.DownStops[req.SourceFloor] = true
	} else {
		// Currently on the exact floor, skip direct routing and immediately target destination
		if req.DestinationFloor > e.CurrentFloor {
			e.UpStops[req.DestinationFloor] = true
		} else {
			e.DownStops[req.DestinationFloor] = true
		}
	}

	// Register destination floor mathematically
	if req.DestinationFloor > req.SourceFloor {
		e.UpStops[req.DestinationFloor] = true
	} else if req.DestinationFloor < req.SourceFloor {
		e.DownStops[req.DestinationFloor] = true
	}

	// Wake up elevator direction if idle
	if e.CurrentDirection == IDLE {
		if req.SourceFloor > e.CurrentFloor || req.DestinationFloor > e.CurrentFloor {
			e.CurrentDirection = UP
		} else if req.SourceFloor < e.CurrentFloor || req.DestinationFloor < e.CurrentFloor {
			e.CurrentDirection = DOWN
		}
	}

	return nil
}

func (e *Elevator) GetDistanceIfAssigned(req *Request) int {
	e.mu.RLock()
	defer e.mu.RUnlock()

	// If capacity is breached, immediately discard it mathematically
	if e.CurrentLoad+req.Passengers > e.Capacity {
		return math.MaxInt32
	}

	dist := int(math.Abs(float64(e.CurrentFloor - req.SourceFloor)))
	
	// Evaluate directional efficiency (Optimal matching algorithm)
	if e.CurrentDirection == IDLE {
		return dist
	}
	if e.CurrentDirection == UP && req.SourceFloor >= e.CurrentFloor {
		return dist
	}
	if e.CurrentDirection == DOWN && req.SourceFloor <= e.CurrentFloor {
		return dist
	}

	// Returning extremely high distance penalty if moving completely opposite to the request
	return dist + 1000 
}

func (e *Elevator) processCurrentFloor() {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.CurrentDirection == UP {
		if e.UpStops[e.CurrentFloor] {
			fmt.Printf("Elevator %s 🛑 STOPPED at floor %d (UP)\n", e.ID, e.CurrentFloor)
			delete(e.UpStops, e.CurrentFloor)
			e.dropOffPassengers()
		}
	} else if e.CurrentDirection == DOWN {
		if e.DownStops[e.CurrentFloor] {
			fmt.Printf("Elevator %s 🛑 STOPPED at floor %d (DOWN)\n", e.ID, e.CurrentFloor)
			delete(e.DownStops, e.CurrentFloor)
			e.dropOffPassengers()
		}
	}
}

func (e *Elevator) dropOffPassengers() {
	var remaining []*Request
	for _, req := range e.Requests {
		if req.DestinationFloor == e.CurrentFloor {
			e.CurrentLoad -= req.Passengers
			fmt.Printf("   -> %d passengers ALIGHTED at floor %d from elevator %s\n", req.Passengers, e.CurrentFloor, e.ID)
		} else {
			remaining = append(remaining, req)
		}
	}
	e.Requests = remaining
}

func (e *Elevator) getNextDirection() Direction {
	e.mu.RLock()
	defer e.mu.RUnlock()

	if e.CurrentDirection == UP {
		// Keep going UP if there are stops above
		for floor := range e.UpStops {
			if floor > e.CurrentFloor {
				return UP
			}
		}
		// Otherwise switch downward bounds
		if len(e.DownStops) > 0 {
			return DOWN
		}
	} else if e.CurrentDirection == DOWN {
		// Keep going DOWN if there are stops below
		for floor := range e.DownStops {
			if floor < e.CurrentFloor {
				return DOWN
			}
		}
		// Otherwise switch upward bounds
		if len(e.UpStops) > 0 {
			return UP
		}
	}

	// IDLE recovery sweep
	if len(e.UpStops) > 0 {
		return UP
	} else if len(e.DownStops) > 0 {
		return DOWN
	}

	return IDLE
}

func (e *Elevator) run() {
	for {
		time.Sleep(100 * time.Millisecond) // Simulated travel time between floors

		nextDir := e.getNextDirection()

		e.mu.Lock()
		e.CurrentDirection = nextDir
		e.mu.Unlock()

		if e.CurrentDirection == IDLE {
			continue // Spin gracefully
		}

		if e.CurrentDirection == UP {
			e.mu.Lock()
			e.CurrentFloor++
			e.mu.Unlock()
		} else if e.CurrentDirection == DOWN {
			e.mu.Lock()
			e.CurrentFloor--
			e.mu.Unlock()
		}

		// Evaluate limits internally per floor change
		e.processCurrentFloor()
	}
}

File: elevator-system/java/Elevator.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Represents a single concurrent elevator, managing its state and processing requests.
 */
public class Elevator implements Runnable {
    private final String id;
    private int currentFloor;
    private Direction currentDirection;
    private final int capacity;
    private int currentLoad;

    private final Map<Integer, Boolean> upStops;
    private final Map<Integer, Boolean> downStops;
    private List<Request> requests;
    
    private volatile boolean running;

    public Elevator(String id, int capacity) {
        this.id = id;
        this.currentFloor = 0;
        this.currentDirection = Direction.IDLE;
        this.capacity = capacity;
        this.currentLoad = 0;
        this.upStops = new HashMap<>();
        this.downStops = new HashMap<>();
        this.requests = new ArrayList<>();
        this.running = true;
    }

    public String getId() {
        return id;
    }

    public synchronized Direction getCurrentDirection() {
        return currentDirection;
    }
    
    public void stopElevator() {
        running = false;
    }

    /**
     * Synchronized evaluation to determine if this elevator is suitable for a request concurrently.
     * By making this method synchronized, we ensure the dispatcher reads an accurate snapshot
     * of currentLoad, currentFloor, and currentDirection without encountering race conditions.
     */
    public synchronized int getDistanceIfAssigned(Request req) {
        if (currentLoad + req.getPassengers() > capacity) {
            return Integer.MAX_VALUE;
        }

        int dist = Math.abs(currentFloor - req.getSourceFloor());

        if (currentDirection == Direction.IDLE) {
            return dist;
        }
        if (currentDirection == Direction.UP && req.getSourceFloor() >= currentFloor) {
            return dist;
        }
        if (currentDirection == Direction.DOWN && req.getSourceFloor() <= currentFloor) {
            return dist;
        }

        return dist + 1000;
    }

    /**
     * Adds a request to the elevator's internal queue in a thread-safe manner.
     * After updating the required stops, it uses notifyAll() to wake up the elevator's run()
     * loop if it was currently waiting (sleeping) in an IDLE state.
     */
    public synchronized void addRequest(Request req) throws Exception {
        if (currentLoad + req.getPassengers() > capacity) {
            throw new Exception("elevator " + id + " capacity exceeded");
        }

        requests.add(req);
        currentLoad += req.getPassengers();

        if (req.getSourceFloor() > currentFloor) {
            upStops.put(req.getSourceFloor(), true);
        } else if (req.getSourceFloor() < currentFloor) {
            downStops.put(req.getSourceFloor(), true);
        } else {
            if (req.getDestinationFloor() > currentFloor) {
                upStops.put(req.getDestinationFloor(), true);
            } else {
                downStops.put(req.getDestinationFloor(), true);
            }
        }

        if (req.getDestinationFloor() > req.getSourceFloor()) {
            upStops.put(req.getDestinationFloor(), true);
        } else if (req.getDestinationFloor() < req.getSourceFloor()) {
            downStops.put(req.getDestinationFloor(), true);
        }

        if (currentDirection == Direction.IDLE) {
            if (req.getSourceFloor() > currentFloor || req.getDestinationFloor() > currentFloor) {
                currentDirection = Direction.UP;
            } else if (req.getSourceFloor() < currentFloor || req.getDestinationFloor() < currentFloor) {
                currentDirection = Direction.DOWN;
            }
        }
        
        // Notify the elevator thread that new work arrived
        notifyAll();
    }

    /**
     * The main execution loop for the Elevator thread, running asynchronously.
     */
    @Override
    public void run() {
        while (running) {
            synchronized (this) {
                // If there's no work, pause the thread execution using wait() to save CPU cycles.
                // The thread will remain asleep until addRequest() invokes notifyAll().
                if (currentDirection == Direction.IDLE && upStops.isEmpty() && downStops.isEmpty()) {
                    try {
                        wait(500); // Wait for new requests to arrive, timeout to check 'running' flag periodically
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                    continue; // Re-evaluate after wake up
                }

                Direction nextDir = getNextDirection();
                currentDirection = nextDir;

                if (currentDirection != Direction.IDLE) {
                    if (currentDirection == Direction.UP) {
                        currentFloor++;
                    } else if (currentDirection == Direction.DOWN) {
                        currentFloor--;
                    }
                }
                
                processCurrentFloor();
            }

            // Simulate the physical time it takes to move between floors outside the synchronized block
            // to allow dispatchers to add requests while it moves.
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
        System.out.printf("   [SYSTEM] Elevator %s has safely shut down.\n", id);
    }

    private synchronized void processCurrentFloor() {
        if (currentDirection == Direction.UP) {
            if (upStops.containsKey(currentFloor) && upStops.get(currentFloor)) {
                System.out.printf("   -> Elevator %s \uD83D\uDED1 STOPPED at floor %d (UP)\n", id, currentFloor);
                upStops.remove(currentFloor);
                dropOffPassengers();
            }
        } else if (currentDirection == Direction.DOWN) {
            if (downStops.containsKey(currentFloor) && downStops.get(currentFloor)) {
                System.out.printf("   -> Elevator %s \uD83D\uDED1 STOPPED at floor %d (DOWN)\n", id, currentFloor);
                downStops.remove(currentFloor);
                dropOffPassengers();
            }
        }
    }

    private synchronized void dropOffPassengers() {
        List<Request> remaining = new ArrayList<>();
        for (Request req : requests) {
            if (req.getDestinationFloor() == currentFloor) {
                currentLoad -= req.getPassengers();
                System.out.printf("      -> %d passengers ALIGHTED at floor %d from elevator %s\n", 
                        req.getPassengers(), currentFloor, id);
            } else {
                remaining.add(req);
            }
        }
        requests = remaining;
    }

    private synchronized Direction getNextDirection() {
        if (currentDirection == Direction.UP) {
            for (int floor : upStops.keySet()) {
                if (floor > currentFloor) {
                    return Direction.UP;
                }
            }
            if (!downStops.isEmpty()) {
                return Direction.DOWN;
            }
        } else if (currentDirection == Direction.DOWN) {
            for (int floor : downStops.keySet()) {
                if (floor < currentFloor) {
                    return Direction.DOWN;
                }
            }
            if (!upStops.isEmpty()) {
                return Direction.UP;
            }
        }

        if (!upStops.isEmpty()) {
            return Direction.UP;
        } else if (!downStops.isEmpty()) {
            return Direction.DOWN;
        }

        return Direction.IDLE;
    }
}
